### PR TITLE
Reduce likelihood of Postgres table scanning `state_groups_state`.

### PR DESCRIPTION
**Pull request by erikjohnston**
The postgres statistics collector sometimes massively underestimates the
number of distinct state groups are in the `state_groups_state`, which
can cause postgres to use table scans for queries for multiple state
groups.

We fix this by manually setting `n_distinct` on the column.

Note that the value `-0.02` was taken by looking at jki.re's database.

Fixes https://github.com/matrix-org/synapse/issues/5064

### CODE REVIEW

File: synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+--
+-- To work around this we can manually tell postgres the number of distint state
+-- groups there are by setting `n_distinct` (a negative value here is the number
+-- of distinct values divided by the number of rows, so -0.02 means on average
+-- there are 50 rows per disinct value). We don't need a particularly
+-- accurate number here, as a) we just want it to always use index scans and b)
+-- our estimate is going to be better than the one made by the statistics
+-- collector.
+
+ALTER TABLE state_groups_state ALTER COLUMN state_group SET (n_dinstict = -0.02)
```
Comment (Half-Shot) [inline synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres:29]: itym
```suggestion
ALTER TABLE state_groups_state ALTER COLUMN state_group SET (n_distinct = -0.02)
```

Comment (erikjohnston) [reply synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres]: Thanks! I was confused why this didn't error, but it turns out if you forget the semicolon on the end we don't even run the command...

File: synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+-- of distinct values divided by the number of rows, so -0.02 means on average
+-- there are 50 rows per disinct value). We don't need a particularly
+-- accurate number here, as a) we just want it to always use index scans and b)
+-- our estimate is going to be better than the one made by the statistics
+-- collector.
+
+ALTER TABLE state_groups_state ALTER COLUMN state_group SET (n_distinct = -0.02);
+
+-- Ideally we'd do an `ANALYZE state_groups_state (state_group)` here so that
+-- the above gets picked up immediately, but that can take a bit of time so we
```
Comment (grinapo) [inline synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres:32]: As a sidenote: on my 72GB sized db the ANALYZE above it took about 1500 ms.

Comment (MTRNord) [reply synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres:32]: For my db ist was around 2.5s but the result of analyze caused it to be noticeable slower than with the ALTER TABLE.

File: synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+-- By default the postgres statistics collector massively underestimates the
+-- number of distinct state groups are in the `state_groups_state`, which can
+-- cause postgres to use table scans for queries for multiple state groups.
+--
+-- To work around this we can manually tell postgres the number of distint state
```
Comment (richvdh) [inline synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres:21]:
```suggestion
-- To work around this we can manually tell postgres the number of distinct state
```

File: synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+-- groups there are by setting `n_distinct` (a negative value here is the number
+-- of distinct values divided by the number of rows, so -0.02 means on average
+-- there are 50 rows per disinct value). We don't need a particularly
```
Comment (richvdh) [inline synapse/storage/schema/main/delta/61/02state_groups_state_n_distinct.sql.postgres:24]:
```suggestion
-- there are 50 rows per distinct value). We don't need a particularly
```

**Review by richvdh — APPROVED** [review]

**Pull request merged**